src/Reflex/Host/Basic.hs

```diff
@@ -87,109 +86,142 @@ instance ReflexHost t => MonadSample t (BasicGuest t m) where
+    traverseIntMapWithKeyWithAdjust (\k v -> unBasicGuest (f k v)) dm0 dm'
+
   {-# INLINABLE traverseDMapWithKeyWithAdjust #-}
-  traverseDMapWithKeyWithAdjust f dm0 dm' = do
-    BasicGuest $ traverseDMapWithKeyWithAdjust (\k v -> unBasicGuest (f k v)) dm0 dm'
-  {-# INLINABLE traverseDMapWithKeyWithAdjustWithMove #-}
-  traverseDMapWithKeyWithAdjustWithMove f dm0 dm' = do
-    BasicGuest $ traverseDMapWithKeyWithAdjustWithMove (\k v -> unBasicGuest (f k v)) dm0 dm'
+  traverseDMapWithKeyWithAdjust f dm0 dm' = BasicGuest $
+    traverseDMapWithKeyWithAdjust (\k v -> unBasicGuest (f k v)) dm0 dm'
 
+  {-# INLINABLE traverseDMapWithKeyWithAdjustWithMove #-}
+  traverseDMapWithKeyWithAdjustWithMove f dm0 dm' = BasicGuest $
+    traverseDMapWithKeyWithAdjustWithMove (\k v -> unBasicGuest (f k v)) dm0 dm'
 
-instance (ReflexHost t) => NotReady t (BasicGuest t m) where
+instance ReflexHost t => NotReady t (BasicGuest t m) where
   {-# INLINABLE notReadyUntil #-}
   notReadyUntil _ = pure ()
+
   {-# INLINABLE notReady #-}
   notReady = pure ()
 
--- | Run a 'BasicGuest' without a quit 'Event'
-basicHostForever :: (forall t m. BasicGuestConstraints t m => BasicGuest t m a)
-                 -> IO a
-basicHostForever guest = basicHostWithQuit $ (\x -> (x, never)) <$> guest
+-- | Run a 'BasicGuest' without a quit 'Event'.
+basicHostForever
+  :: (forall t m. BasicGuestConstraints t m => BasicGuest t m a)
+  -> IO a
+basicHostForever guest = basicHostWithQuit $ guest <&> (,never)
 
--- | Run a 'BasicGuest'
+-- | Run a 'BasicGuest'.
 --
 -- The program will exit when the 'Event' returned by the 'BasicGuest' fires
-basicHostWithQuit :: (forall t m. BasicGuestConstraints t m => BasicGuest t m (a, Event t ()))
-                  -> IO a
+basicHostWithQuit
+  :: (forall t m. BasicGuestConstraints t m => BasicGuest t m (a, Event t ()))
+  -> IO a
 basicHostWithQuit (BasicGuest guest) = do
-  events <- liftIO newChan
+  performEventChan <- newChan
+  (postBuild, postBuildTriggerRef) <- runSpiderHost newEventWithTriggerRef
 
-  rHasQuit <- liftIO $ newIORef False
+  -- Unpack the guest
+  ((a, eQuit), FireCommand fire) <- runSpiderHost . hostPerformEventT $
+    runTriggerEventT (runPostBuildT guest postBuild) performEventChan
 
-  ((a, eQuit), FireCommand fire) <- liftIO $ runSpiderHost $ do
-    (((a, eQuit), postBuildTriggerRef), fc@(FireCommand fire)) <- hostPerformEventT $ do
-      (postBuild, postBuildTriggerRef) <- newEventWithTriggerRef
-      pae <- runTriggerEventT (runPostBuildT guest postBuild) events
-      pure (pae, postBuildTriggerRef)
-
-    hQuit <- subscribeEvent eQuit
-
-    mPostBuildTrigger <- readRef postBuildTriggerRef
-    forM_ mPostBuildTrigger $ \postBuildTrigger -> do
-      lmQuit <- fire [postBuildTrigger :=> Identity ()] $ readEvent hQuit >>= sequence
-      when (any isJust lmQuit) $
-        liftIO $ writeIORef rHasQuit True
+  hQuit <- runSpiderHost $ subscribeEvent eQuit
+  rHasQuit <- newRef False -- When to shut down
```

Comment:
If this were all to be folded into a single call to `runSpiderHost`, this would probably move to before the call to `runSpiderHost . hostPerformEventT`.